components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt

```diff
@@ -155,6 +159,17 @@ internal class WorkManagerSyncDispatcher(
             notifyObservers { onStarted() }
             isSyncActive = true
         }
+        crashReporter?.recordCrashBreadcrumb(
+            Breadcrumb(
+                category = this::class.java.simpleName,
+                message = "syncing-state-change",
+                data = mapOf(
+                    "active" to isSyncActive.toString(),
+                    "engines" to supportedEngines.map { it.nativeName }.joinToString(","),
```

Comment:
`engines` here will be all engines supported by this client, not the engines that are being synced. I don't think that was your intention though?

I think the way to go about this is to expand `SyncStatusObserver#onStarted` to receive a list of engines being synced.

To do that, at the point where we detect a status change we need to have info about engines for which the sync was triggered. Currently that's passed into the worker as part of a `WorkRequest`'s data. However, when we observe `WorkManager` for worker status changes, we get `WorkInfo` instances instead - which do not have input data passed into the workers. But they do have `tags` available. So if we set engines to sync and reason as tags of the `WorkRequests` we're constructing, it should all work.

`WorkRequest` instances are constructed in `regularSyncWorkRequest` and `periodicSyncWorkRequest`. 

You can also pass along as a tag and record `reason` for syncing in a breadcrumb, as well - e.g. is this a manual sync triggered by a user, a startup sync, a periodic sync, etc.

Then, we can register an observer on an instance of `WorkManagerSyncDispatcher` that will be responsible for recording breadcrumbs. That in itself can either happen somewhere near this code, or in the app. I vaguely prefer to do this in the app, but A-C works as well. We already have an established pattern of recording breadcrumbs elsewhere in A-C, so I guess that's fine.

Reply:
Thanks Grisha. I'm struggling with this part:

> I think the way to go about this is to expand `SyncStatusObserver#onStarted` to receive a list of engines being synced.

That makes sense - but I'm struggling with working out how to get that list of engines. In my patch, I'm calling `recordCrashBreadcrumb` just 2 lines after where we call `onStarted()` - and I can't work out how to get that in that function.

> To do that, at the point where we detect a status change we need to have info about engines for which the sync was triggered. Currently that's passed into the worker as part of a `WorkRequest`'s data.

Yep.

> However, when we observe `WorkManager` for worker status changes, we get `WorkInfo` instances instead - which do not have input data passed into the workers. But they do have `tags` available.

They do have tags, but best I can tell, there's no "associated data" - ie, it's literally just a string tag. I can query all tags, and maybe you meant I could add, say, "engine:logins" and "engine:places", then re-constitute the list of engines from these tags, but that seems a little odd.

The WorkManager does know the list of engines, but best I can tell, the observer has no reference to that.

So I'm a bit lost how to weave that data through to where it needs to go. Any clues?

Reply:
>  can query all tags, and maybe you meant I could add, say, "engine:logins" and "engine:places", then re-constitute the list of engines from these tags, but that seems a little odd.

That's what I meant, yeah, and couldn't find another way to do this without, say, some extra global state. We're limited by what can be stuffed into a `WorkInfo`, and `tags: Set<String>` seems like the best approach.

I guess an alternative to this would be some global state where you record individual instances of a sync (each gets a UUID), and which engines are being synced. Then in a `WorkRequest` tag, we can stuff that sync's UUID, and look up the engine information from the global state.

I'm not really sure what's worse, tbh.

Although, at the end of the day, if all you're doing is recording a breadcrumb that looks something like:
```
message = "syncing-state-change",
data = mapOf("active" to 1, "engines" to "1,3")
```

... then stuffing a tag `engines:1,3` into a WorkRequest perhaps isn't too bad. You don't _really_ need the proper type, and if you do, you can easily get it.